recirq/qml_lfe/learn_dynamics_c.py

```diff
@@ -0,0 +1,191 @@
+
+
+flags.DEFINE_integer("n", None, "System size (total qubits == 2 * n).")
+flags.DEFINE_integer("depth", None, "Circuit depth (block-wise).")
+flags.DEFINE_integer(
+    "n_data",
+    20,
+    "Number of circuits generated for each class (total circuits == 2 * n_data).",
+)
+
+flags.DEFINE_integer(
+    "batch_size",
+    5,
+    "Number of circuits to send over the wire per batch (value does not affect results).",
+)
+
+flags.DEFINE_integer(
+    "n_shots", 2000, "Number of measurements to draw from each individual circuit."
+)
+
+flags.DEFINE_string(
+    "save_dir", "./data", "Path to save experiment data (must already exist)."
+)
+
+flags.DEFINE_bool("use_engine", False, "Whether or not to use quantum engine.")
+
+
+def flatten_circuit(circuit: cirq.Circuit) -> cirq.Circuit:
+    """Pack operations in circuit to the left as far as possible."""
+    return cirq.Circuit([op for mom in circuit for op in mom])
+
+
+def build_circuit(
+    qubit_pairs: List[List[cirq.Qid]], use_tsym: bool, depth: int
+) -> cirq.Circuit:
+    """Generate 1D tsym or scrambling circuit with given depth on qubit_pairs."""
+    inter_gen = circuit_blocks.scrambling_block
+    if use_tsym:
+        # Use tsym gates.
+        inter_gen = circuit_blocks.tsym_block
+
+    # Random source for circuits.
+    random_source = np.random.uniform(0, 4, size=(depth * len(qubit_pairs), 2))
+
+    ret_circuit = circuit_blocks.block_1d_circuit(
+        [qubits[0] for qubits in qubit_pairs], depth, inter_gen, random_source
+    )
+    ret_circuit += [cirq.S(qubits[0]) for qubits in qubit_pairs]
+    ret_circuit += [cirq.H(qubits[0]) for qubits in qubit_pairs]
+
+    # Merge single qubit gates together and add measurements.
+    cirq.merge_single_qubit_gates_into_phxz(ret_circuit)
+    cirq.DropEmptyMoments().optimize_circuit(circuit=ret_circuit)
+    ret_circuit = flatten_circuit(ret_circuit)
+
+    for i, qubit in enumerate([qubits[0] for qubits in qubit_pairs]):
+        ret_circuit += cirq.measure(qubit, key=f"q{i}")
+
+    cirq.SynchronizeTerminalMeasurements().optimize_circuit(circuit=ret_circuit)
+    logging.debug(
+        f"Generated a new circuit w/ tsym={use_tsym} and depth {len(ret_circuit)}"
+    )
+    return ret_circuit
+
+
+def _engine_sim_workaround(batch: List[cirq.Circuit]) -> List[cirq.Result]:
+    """Use either simulator or engine."""
+    if FLAGS.use_engine:
+        project_id = os.environ["GOOGLE_CLOUD_PROJECT"]
+        engine = cirq_google.Engine(project_id=project_id)
+        job = engine.run_batch(
+            programs=batch,
+            gate_set=cirq_google.SYC_GATESET,
+            processor_ids=["weber"],
+            repetitions=FLAGS.n_shots,
+        )
+        return job.results()
+
+    sim = cirq.Simulator()
+    return [sim.run(circuit, repetitions=FLAGS.n_shots) for circuit in batch]
+
+
+def main(_):
+    logging.info("Beginning conventional circuit generation.")
+    # Choose system pairs so that they are consecutive neighbors.
```

Comment:
Maybe add to the comment what this layout is based on (Sycamore54 on Weber?)

Reply:
Done.